### ISSUE TITLE
Bug report: Indexing a list of floats in a painless script seems to access the elements in sorted order

### ISSUE THREAD

**Issue by alexklibisz** [elastic/elasticsearch#49695]
**Elasticsearch version** (`bin/elasticsearch --version`): 

```
# bin/elasticsearch -version
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Version: 7.4.0, Build: default/docker/22e1767283e61a198cb4db791ea66e3f11ab9910/2019-09-27T08:36:48.569419Z, JVM: 13
```

**Plugins installed**: No plugins installed

**JVM version** (`java -version`):

```
openjdk version "13" 2019-09-17
OpenJDK Runtime Environment AdoptOpenJDK (build 13+33)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 13+33, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system):

```
Linux 27024d91b217 5.0.0-36-generic #39-Ubuntu SMP Tue Nov 12 09:46:06 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:

I am running the elasticsearch and kibana 7.4.0 docker containers locally. 

I am storing a list of floats (i.e. a "vector") in each document. I have a script score query with a painless script that loops through the vector and does some computation to return a score. I was getting some incorrect results, and after some experimenting I found that indexing into the list seems to access the elements in sorted order, not in the order in which they were indexed. Perhaps this is intended behavior, but it's certainly surprising. I also wonder whether sorting the vectors could make indexing slower, but this is a lesser concern.

**Steps to reproduce**:

Here are the exact requests you should run to reproduce the issue. I think it's easiest to run them from the kibana console. You could just copy them into the kibana console, highlight them, and run them all.

```
# Delete and re-create index.
DELETE vectors
PUT vectors

# Store a vector already in sorted order
PUT vectors/_doc/sorted
{
  "values": [0.123,0.234,0.345]
}

# Store a vector not in sorted order.
PUT vectors/_doc/unsorted
{
  "values": [0.333,0.111,0.222]
}

# Refresh the index before searching
POST vectors/_refresh

# Run a scripted search that just returns the 0th element.
GET vectors/_search
{
  "query": {
    "script_score": {
      "query": {
        "exists": {
          "field": "values"
        }
      },
      "script": {
        "source": "return doc['values'][0];",
        "lang": "painless"
      }
    }
  }
}
```

Here is the result for the last search request. Notice:
- good: each list of values is returned in the correct order
- good: the score for the sorted document is 0.123, which is the 0th (and smallest) element
- **bad**: the score for the unsorted document is 0.111, the smallest element, but not the 0th. It should have been 0.333.

```
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : 0.123,
    "hits" : [
      {
        "_index" : "vectors",
        "_type" : "_doc",
        "_id" : "sorted",
        "_score" : 0.123,
        "_source" : {
          "values" : [
            0.123,
            0.234,
            0.345
          ]
        }
      },
      {
        "_index" : "vectors",
        "_type" : "_doc",
        "_id" : "unsorted",
        "_score" : 0.111,
        "_source" : {
          "values" : [
            0.333,
            0.111,
            0.222
          ]
        }
      }
    ]
  }
}
```

**Provide logs (if relevant)**:

**Comment by alexklibisz** [comment]
After a bit more googling, I found that you can also access the original source using `params._source`.
So now this version of the script is working:

```
# Run a scripted search that just returns the 0th element.
GET vectors/_search
{
  "query": {
    "script_score": {
      "query": {
        "exists": {
          "field": "values"
        }
      },
      "script": {
        "source": "return params._source['values'][0];",
        "lang": "painless"
      }
    }
  }
}
```

These SO questions were helpful:
- https://stackoverflow.com/questions/41972733/elasticsearch-array-field-is-sorted-when-accessed-within-painless-scripting-lang?rq=1
- https://stackoverflow.com/questions/55219970/es-painlessdifference-between-doc-and-params-source?rq=1

It would be great to get confirmation that this is the correct way to solve this problem. Thanks

**Comment by jimczi** [comment]
Multi-valued field are sorted at index time in the doc-values so painless is not responsible of this ordering. This is intended but we should document this more clearly which is why I am marking this issue with the docs tag. 
Regarding the workaround, it is not recommended to access the source directly since your query will need to retrieve the entire json to extract a single value on all documents. You could instead use a separate field that records the first value and use the doc values of this field in your script score ?

**Comment by alexklibisz** [comment]
> Multi-valued field are sorted at index time in the doc-values so painless is not responsible of this ordering. This is intended but we should document this more clearly which is why I am marking this issue with the docs tag.
> Regarding the workaround, it is not recommended to access the source directly since your query will need to retrieve the entire json to extract a single value on all documents. You could instead use a separate field that records the first value and use the doc values of this field in your script score ?

I appreciate the quick response. I don't think I understand your recommendation. Could you show an example of the document JSON for the pattern you proposed?

**Comment by alexklibisz** [comment]
Here are a couple other ways I could handle this:

- Store the index alongside each element:
```
{ "values": { "0": 0.333, "1": 0.111, "2": 0.222 } }
```
I'm expecting to always have dense vectors, so this seems like it would waste a lot of space.

- Store the vector as a comma-separated string:
```
{ "values": "0.333,0.111,0.222" }
```
I don't like this either since the added string parsing will likely slow down the script.

Is there any way to just disable the index-time sorting?

**Comment by jimczi** [comment]
> Is there any way to just disable the index-time sorting?

We have a [dense vector](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html) field type so you should take a look there. Can you describe the scoring that you apply on your vectors ? We have some [predefined functions](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-script-score-query.html#vector-functions) that you could use.

**Comment by alexklibisz** [comment]
> > Is there any way to just disable the index-time sorting?
> 
> We have a [dense vector](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html) field type so you should take a look there. Can you describe the scoring that you apply on your vectors ? We have some [predefined functions](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-script-score-query.html#vector-functions) that you could use.

I'm computing distance functions, but not all of the functions I want are provided. Also, I'm confused by the "X-Pack" tags on those features, so I've avoided them. I assumed it means I have to have to have the commercial version of ES to use them.

**Comment by alexklibisz** [comment]
I looked more at the dense vector type. There are a few issues for my use case:
1. It only supports storing floats. There are cases where I need to store boolean vectors as well.
2. It doesn't support accessing values from a custom script. The `.get` method throws an exception when trying to access an index.
3. I noticed sparse vectors were recently deprecated due to lack of adoption. Could dense vectors also be deprecated?

I acknwoledge my use-case is a bit niche, and I don't want to hog resources. It would be best if there's a way to disable the index-time sorting so I can access the doc-values vector in its original order. If not, I'll just continue accessing via `params._source`, and if it becomes a performance bottleneck I'll revisit the issue.

**Comment by alexklibisz** [comment]
Closing as I was able to work around this by creating a mapper plugin to store my double and boolean vectors with a proper `get(i)` method to access the elements in their original order. I'm planning to open source this work so I'll try to remember to post a link on this issue once I do. 

It would still be great to get some additional docs describing the original quirk about accessing elements in lists from painless scripts. Thanks for your help @jimczi.
